theano/tensor/signal/pool.py

```diff
@@ -2438,3 +2438,185 @@ def c_code(self, node, name, inp, out, sub):
+
+    def infer_shape(self, node, in_shapes):
+        return [in_shapes[0]]
+
+    def get_op_params(self):
+        return [('POOLED_HEIGHT', str(self.pooled_h)),
+                ('POOLED_WIDTH', str(self.pooled_w)),
+                ('SPATIAL_SCALE', str(self.spatial_scale))]
+
+    def perform(self, node, inp, out):
+        image_data, roi, argmax, out_grad = inp
+        gx, = out
+        num_roi = roi.shape[0]
+        spatial_scale = self.spatial_scale
+        pool_height = self.pooled_h
+        pool_width = self.pooled_w
+        batch_size = image_data.shape[0]
+        n_channels = image_data.shape[1]
+        image_height = image_data.shape[2]
+        image_width = image_data.shape[3]
+        assert image_data.ndim == 4
+        assert roi.ndim == 2
+        gx[0] = numpy.zeros(image_data.shape)
+        gxx = gx[0]
+        for b_in in range(batch_size):
+            for i in range(num_roi):
+                x_start = numpy.floor((roi[i, 1] * spatial_scale) + 0.5)
+                y_start = numpy.floor((roi[i, 2] * spatial_scale) + 0.5)
+                x_end = numpy.floor((roi[i, 3] * spatial_scale) + 0.5)
+                y_end = numpy.floor((roi[i, 4] * spatial_scale) + 0.5)
+
+                roi_height = max(y_end - y_start + 1, 1)
+                roi_width = max(x_end - x_start + 1, 1)
+                row_length = roi_width / pool_width
+                col_length = roi_height / pool_height
+
+                for cn in range(n_channels):
+                    channel_grad = out_grad[b_in, i, cn]
+                    gxxx = gxx[b_in][cn]
+                    gxxx.shape = (image_height * image_width, )
+                    for jy in range(image_height):
+                        for ix in range(image_width):
+                            in_roi = (ix >= x_start and ix <= x_end and jy >= y_start and jy <= y_end)
+                            if not in_roi:
+                                continue
+                            bottom_index = (jy * image_width) + ix
+                            x1 = int(numpy.floor((ix - x_start) / row_length))
+                            x2 = int(numpy.ceil((ix - x_start + 1) / row_length))
+                            y1 = int(numpy.floor((jy - y_start) / col_length))
+                            y2 = int(numpy.ceil((jy - y_start + 1) / col_length))
+
+                            interest_region = argmax[b_in, i, cn, x1 + int(pool_width) * y1:x2 + int(pool_width) * y2]
+
+                            mp_index = numpy.where(interest_region == bottom_index)[0]
+                            if mp_index.size == 0:
+                                continue
+
+                            # incrementing cause it was extracted from sliced array
+                            mp_index = numpy.add(mp_index, x1 + (pool_width * y1))
+                            gxxx[bottom_index] += channel_grad[mp_index].sum()
+
+    def grad(self, inp, grads):
+        disc = [tensor.zeros_like(i) for i in inp]
```

Comment:
I'm pretty sure this isn't right. Can you just remove that method instead? Currently it isn't tested too. So if you want to get that in Theano for the next release, just remove it. If needed it can be added later with tests.

Reply:
I remember in some comment where @f0k had told me that this `grad` method is for computing the gradient of gradient. Since it's something that's not necessary, I had kept it as zeroes. Shall I change it to not implemented rather?